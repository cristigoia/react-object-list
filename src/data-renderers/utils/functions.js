/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
