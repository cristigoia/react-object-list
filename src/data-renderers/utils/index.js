export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
