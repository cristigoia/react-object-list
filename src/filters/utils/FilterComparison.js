import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
