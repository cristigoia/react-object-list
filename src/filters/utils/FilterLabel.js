import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
