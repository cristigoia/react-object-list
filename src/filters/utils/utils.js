const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
