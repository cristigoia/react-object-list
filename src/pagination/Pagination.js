import React from 'react'
import PropTypes from 'prop-types'
import Page from './Page'

/**
 * Component used to display/control pagination
 */
export default class Pagination extends React.Component {
  static propTypes = {
    // Initial page of results
    page: PropTypes.number,
    // Number of items to display per page
    perPage: PropTypes.number,
    // Maximum number of pages to display
    maxPages: PropTypes.number,
    // Total number of results
    count: PropTypes.number,
    // Function to call to change page
    goToPage: PropTypes.func.isRequired,
    // True if results are loading
    loading: PropTypes.bool,
    // Icon to display if results are loading
    LoadingIcon: PropTypes.func,
    // plural name for items displayed in the list
    itemPluralName: PropTypes.string,
  }

  static defaultProps = {
    page: 1,
    perPage: 50,
    maxPages: 10,
  }

  render() {
    const {page, maxPages, count, perPage, loading, LoadingIcon, goToPage, itemPluralName} = this.props
    const pages = []
    let totalPages = 0
    const minPage = Math.min(
      Math.ceil(page - (maxPages / 2)),
      Math.ceil(count / perPage) - maxPages + 1
    )
    for (let position = 0; position < count; position += perPage) {
      const pageIndex = Math.floor(position / perPage) + 1
      totalPages++
      if (pageIndex < minPage) { continue }
      if (pages.length >= maxPages) { continue }
      pages.push(
        <Page
          key={`page-${pageIndex}`}
          page={pageIndex}
          active={pageIndex === page}
          goToPage={goToPage}
        />
      )
    }
    if (totalPages === 1) return null
    const prevPage = Math.max(Math.min(totalPages, page - 1), 1)
    const prevLink = (
      <Page
        key="page-prev"
        page={prevPage}
        active={false}
        disabled={page === prevPage}
        label="←&nbsp;Previous"
        goToPage={goToPage}
      />
    )

    const nextPage = Math.max(Math.min(totalPages, page + 1), 1)
    const nextLink = (
      <Page
        key="page-next"
        page={nextPage}
        active={false}
        disabled={page === totalPages}
        label="Next&nbsp;→"
        goToPage={goToPage}
      />
    )
    let totals
    if (!loading) {
      totals = (
        <p className="text-muted">
          <em>
            {`Showing page ${page} of ${Math.max(totalPages, 1)} out of ${count} total ${itemPluralName || 'records'}.`}
          </em>
        </p>
      )
    } else {
      totals = (
        <p className="text-muted">
          <em>
            {LoadingIcon && <LoadingIcon />} Loading {itemPluralName || 'results'} ...
          </em>
        </p>
      )
    }
    return (
      <div className="apilist-pagination">
        <ul className="apilist-pagination__pages">
          {prevLink}
          {pages}
          {nextLink}
        </ul>
        {totals}
      </div>
    )
  }
}
