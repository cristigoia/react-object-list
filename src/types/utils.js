function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
