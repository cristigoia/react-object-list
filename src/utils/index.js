export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'
