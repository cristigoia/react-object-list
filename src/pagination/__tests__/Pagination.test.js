import React from 'react'
import { snapshotTest } from 'utils/tests'
import Pagination from '../Pagination.js'

jest.mock('../Page', () => 'Page')

describe('Pagination', () => {
  const defaultProps = {
    goToPage: jest.fn(),
    page: 1,
    perPage: 10,
    maxPages: 10,
    count: 100,
  }
  describe('Snapshots', () => {
    it('has one page', () => {
      snapshotTest(<Pagination {...defaultProps} count={1} />)
    })
    it('is on first page', () => {
      snapshotTest(<Pagination {...defaultProps} />)
    })
    it('is on last page', () => {
      snapshotTest(<Pagination {...defaultProps} page={10} />)
    })
    it('is loading', () => {
      snapshotTest(<Pagination {...defaultProps} loading />)
    })
    it('has more than maxPages', () => {
      snapshotTest(<Pagination {...defaultProps} maxPages={2} />)
    })
  })
})
