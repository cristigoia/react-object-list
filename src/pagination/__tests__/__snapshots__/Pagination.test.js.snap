// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Snapshots has more than maxPages 1`] = `
<div
  className="objectlist-pagination"
>
  <ul
    className="objectlist-pagination__pages"
  >
    <Page
      active={false}
      disabled={true}
      goToPage={[MockFunction]}
      label="← Previous"
      page={1}
    />
    <Page
      active={true}
      goToPage={[MockFunction]}
      page={1}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={2}
    />
    <Page
      active={false}
      disabled={false}
      goToPage={[MockFunction]}
      label="Next →"
      page={2}
    />
  </ul>
  <p
    className="text-muted"
  >
    <em>
      Showing page 1 of 10 out of 100 total snakes.
    </em>
  </p>
</div>
`;

exports[`Pagination Snapshots has one page 1`] = `null`;

exports[`Pagination Snapshots is loading 1`] = `
<div
  className="objectlist-pagination"
>
  <ul
    className="objectlist-pagination__pages"
  >
    <Page
      active={false}
      disabled={true}
      goToPage={[MockFunction]}
      label="← Previous"
      page={1}
    />
    <Page
      active={true}
      goToPage={[MockFunction]}
      page={1}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={2}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={3}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={4}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={5}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={6}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={7}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={8}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={9}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={10}
    />
    <Page
      active={false}
      disabled={false}
      goToPage={[MockFunction]}
      label="Next →"
      page={2}
    />
  </ul>
  <p
    className="text-muted"
  >
    <em>
       Loading 
      snakes
       ...
    </em>
  </p>
</div>
`;

exports[`Pagination Snapshots is on first page 1`] = `
<div
  className="objectlist-pagination"
>
  <ul
    className="objectlist-pagination__pages"
  >
    <Page
      active={false}
      disabled={true}
      goToPage={[MockFunction]}
      label="← Previous"
      page={1}
    />
    <Page
      active={true}
      goToPage={[MockFunction]}
      page={1}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={2}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={3}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={4}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={5}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={6}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={7}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={8}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={9}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={10}
    />
    <Page
      active={false}
      disabled={false}
      goToPage={[MockFunction]}
      label="Next →"
      page={2}
    />
  </ul>
  <p
    className="text-muted"
  >
    <em>
      Showing page 1 of 10 out of 100 total snakes.
    </em>
  </p>
</div>
`;

exports[`Pagination Snapshots is on last page 1`] = `
<div
  className="objectlist-pagination"
>
  <ul
    className="objectlist-pagination__pages"
  >
    <Page
      active={false}
      disabled={false}
      goToPage={[MockFunction]}
      label="← Previous"
      page={9}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={1}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={2}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={3}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={4}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={5}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={6}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={7}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={8}
    />
    <Page
      active={false}
      goToPage={[MockFunction]}
      page={9}
    />
    <Page
      active={true}
      goToPage={[MockFunction]}
      page={10}
    />
    <Page
      active={false}
      disabled={true}
      goToPage={[MockFunction]}
      label="Next →"
      page={10}
    />
  </ul>
  <p
    className="text-muted"
  >
    <em>
      Showing page 10 of 10 out of 100 total snakes.
    </em>
  </p>
</div>
`;
